File: backend/app/schemas/task.py
"""Task schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.task import TaskPriority, TaskStatus
from app.schemas.user import UserSummary


class TaskBase(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=5000)
    priority: TaskPriority = TaskPriority.MEDIUM


class TaskCreate(TaskBase):
    assigned_to_user_id: int


class TaskUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=5000)
    priority: Optional[TaskPriority] = None
    assigned_to_user_id: Optional[int] = None


class TaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str
    status: TaskStatus
    priority: TaskPriority
    assigned_to_user_id: int
    created_by_user_id: int
    assignee: UserSummary
    creator: UserSummary
    created_at: datetime
    updated_at: datetime


class TaskActionRequest(BaseModel):
    """Optional comment for submit/approve/reject/complete actions."""

    comment: Optional[str] = Field(default=None, max_length=500)

File: backend/src/main/java/com/example/bwms/model/Task.java
// Translated from: backend/app/models/task.py
package com.example.bwms.model;

import jakarta.persistence.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "tasks", indexes = {
        @Index(name = "ix_tasks_status",              columnList = "status"),
        @Index(name = "ix_tasks_priority",            columnList = "priority"),
        @Index(name = "ix_tasks_assigned_to_user_id", columnList = "assigned_to_user_id"),
        @Index(name = "ix_tasks_created_by_user_id",  columnList = "created_by_user_id"),
        @Index(name = "ix_tasks_status_priority",     columnList = "status, priority")
})
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(length = 200, nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String description = "";

    @Column(nullable = false)
    private TaskStatus status = TaskStatus.PENDING;

    @Column(nullable = false)
    private TaskPriority priority = TaskPriority.MEDIUM;

    @Column(name = "assigned_to_user_id", nullable = false)
    private Long assignedToUserId;

    @Column(name = "created_by_user_id", nullable = false)
    private Long createdByUserId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to_user_id", insertable = false, updatable = false)
    private User assignee;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_user_id", insertable = false, updatable = false)
    private User creator;

    @OneToMany(mappedBy = "task", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("createdAt ASC")
    private List<WorkflowEvent> workflowEvents = new ArrayList<>();

    @OneToMany(mappedBy = "task", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("createdAt ASC")
    private List<Approval> approvals = new ArrayList<>();

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private OffsetDateTime updatedAt;

    public Task() {}

    public Long getId()                  { return id; }
    public String getTitle()             { return title; }
    public String getDescription()       { return description; }
    public TaskStatus getStatus()        { return status; }
    public TaskPriority getPriority()    { return priority; }
    public Long getAssignedToUserId()    { return assignedToUserId; }
    public Long getCreatedByUserId()     { return createdByUserId; }
    public User getAssignee()            { return assignee; }
    public User getCreator()             { return creator; }
    public OffsetDateTime getCreatedAt() { return createdAt; }
    public OffsetDateTime getUpdatedAt() { return updatedAt; }

    public void setTitle(String title)               { this.title = title; }
    public void setDescription(String description)   { this.description = description; }
    public void setStatus(TaskStatus status)         { this.status = status; }
    public void setPriority(TaskPriority priority)   { this.priority = priority; }
    public void setAssignedToUserId(Long userId)     { this.assignedToUserId = userId; }
    public void setCreatedByUserId(Long userId)      { this.createdByUserId = userId; }
    public void setCreatedAt(OffsetDateTime createdAt) { this.createdAt = createdAt; }
    public void setUpdatedAt(OffsetDateTime updatedAt) { this.updatedAt = updatedAt; }
}
